File: client-js/lib/profiles.js
/** Contree configuration profiles.
 *
 * The INI files under `$CONTREE_HOME`: `auth.ini`
 * (secrets) merged over `$CONTREE_HOME/cli.ini` (non-secret defaults),
 * where CONTREE_HOME defaults to `$XDG_CONFIG_HOME/contree` and
 * finally `~/.config/contree`. Node-only: the filesystem modules are
 * loaded lazily so browser bundlers never pull them in.
 */

import { DEFAULT_BASE_URL } from "./specInfo.js";

export const AUTH_TYPE_IAM = "iam";
export const AUTH_TYPE_JWT = "jwt";
export const PROFILE_PREFIX = "profile:";
export const DEFAULT_PROFILE = "default";

export class ProfileError extends Error {}

function record() {
  // a null-prototype object: keys come from an external file, and a
  // key like "__proto__" must become plain data, never touch the
  // prototype chain of anything process-wide
  return Object.create(null);
}

/** A minimal INI parser covering the profile-file dialect: `[section]` headers, `key = value` pairs, `#`/`;` comments.
 * `[DEFAULT]` and keys outside any section land in the "" (defaults)
 * entry - exactly like ConfigParser's magic DEFAULT section. Returns
 * a Map of section name to a null-prototype key/value object. */
export function parseIni(text) {
  const sections = new Map([["", record()]]);
  let current = "";
  for (const rawLine of text.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#") || line.startsWith(";")) {
      continue;
    }
    if (line.startsWith("[") && line.endsWith("]")) {
      const name = line.slice(1, -1).trim();
      current = name === "DEFAULT" ? "" : name;
      if (!sections.has(current)) {
        sections.set(current, record());
      }
      continue;
    }
    const eq = line.indexOf("=");
    if (eq < 0) {
      continue;
    }
    const key = line.slice(0, eq).trim().toLowerCase();
    sections.get(current)[key] = line.slice(eq + 1).trim();
  }
  return sections;
}

async function contreeHome() {
  const { homedir } = await import("node:os");
  const { join } = await import("node:path");
  const home = process.env.CONTREE_HOME;
  if (home) {
    return home;
  }
  const xdg = process.env.XDG_CONFIG_HOME || join(homedir(), ".config");
  return join(xdg, "contree");
}

async function readIni(path) {
  const { readFile } = await import("node:fs/promises");
  try {
    return parseIni(await readFile(path, "utf-8"));
  } catch (error) {
    if (error?.code === "ENOENT") {
      return new Map([["", record()]]);
    }
    // a permission or I/O failure must not masquerade as a missing
    // profile - surface it
    throw error;
  }
}

/** Read all profiles; resolves to `{profiles, active}`.
 *
 * `cli.ini` is read first, then `auth.ini` - values from `auth.ini`
 * win on conflicts. `[DEFAULT]` values are inherited by every profile
 * section (ConfigParser semantics, like the Python client).
 */
export async function loadProfiles(path = null) {
  const { dirname, join } = await import("node:path");
  const authFile = path ?? join(await contreeHome(), "auth.ini");
  const merged = new Map([["", record()]]);
  for (const file of [join(dirname(authFile), "cli.ini"), authFile]) {
    for (const [section, values] of await readIni(file)) {
      const target = merged.get(section) ?? record();
      Object.assign(target, values);
      merged.set(section, target);
    }
  }
  const defaults = merged.get("");
  const active = defaults.profile ?? DEFAULT_PROFILE;
  const profiles = record();
  for (const [section, values] of merged) {
    if (!section.startsWith(PROFILE_PREFIX)) {
      continue;
    }
    const name = section.slice(PROFILE_PREFIX.length);
    // ConfigParser exposes DEFAULT values through every section
    const effective = record();
    Object.assign(effective, defaults, values);
    const authType = effective.type ?? AUTH_TYPE_JWT;
    const defaultUrl = authType === AUTH_TYPE_IAM ? DEFAULT_BASE_URL : "";
    profiles[name] = {
      name,
      token: effective.token ?? null,
      url: (effective.url ?? defaultUrl).replace(/\/+$/, ""),
      authType,
      project: effective.project ?? null,
    };
  }
  return { profiles, active };
}

/** Build a profile from the environment, bypassing config files.
 *
 * Reads the standard Contree variables: `CONTREE_TOKEN`
 * (or `NEBIUS_API_KEY`), `CONTREE_URL` and optionally
 * `CONTREE_PROJECT` (or `NEBIUS_AI_PROJECT`). Returns null unless
 * both a token and a URL are present - a non-null result means the
 * environment fully described a profile.
 */
export function fromEnv() {
  const env = globalThis.process?.env ?? {};
  const token = env.CONTREE_TOKEN || env.NEBIUS_API_KEY || null;
  const url = env.CONTREE_URL || null;
  if (!token || !url) {
    return null;
  }
  return {
    name: "environment",
    token,
    url: url.replace(/\/+$/, ""),
    authType: AUTH_TYPE_JWT,
    project: env.CONTREE_PROJECT ?? env.NEBIUS_AI_PROJECT ?? null,
  };
}

/** Resolve the active profile by name.
 *
 * Priority: explicit *name* > `CONTREE_PROFILE` environment variable
 * > the active profile recorded in the config file (`[DEFAULT]`).
 */
export async function resolveProfile(name = null, { path = null } = {}) {
  const { profiles, active } = await loadProfiles(path);
  const selected = name || process.env.CONTREE_PROFILE || active;
  if (!(selected in profiles)) {
    throw new ProfileError(`profile ${JSON.stringify(selected)} not found`);
  }
  return profiles[selected];
}

File: client/contree_client/profiles.py
"""Contree configuration profiles.

The INI files under `$CONTREE_HOME`: `auth.ini`
(secrets) merged over `$CONTREE_HOME/cli.ini` (non-secret defaults),
where `CONTREE_HOME` defaults to `$XDG_CONFIG_HOME/contree` and
finally `~/.config/contree`.
"""

from __future__ import annotations

import configparser
import os
from configparser import ConfigParser
from dataclasses import dataclass, field
from pathlib import Path

from .spec_info import DEFAULT_BASE_URL

AUTH_TYPE_IAM = "iam"
AUTH_TYPE_JWT = "jwt"
PROFILE_PREFIX = "profile:"
DEFAULT_PROFILE = "default"
DEFAULT_IAM_URL = DEFAULT_BASE_URL


class ProfileError(ValueError):
    """The requested profile is missing or incomplete."""


@dataclass(frozen=True)
class Profile:
    """One Contree authentication profile.

    Frozen and comparable; the token never shows up in `repr()`.
    """

    name: str
    url: str
    token: str | None = field(repr=False)
    auth_type: str = AUTH_TYPE_JWT
    project: str | None = None

    def save(self, config: ConfigParser) -> None:
        section = f"{PROFILE_PREFIX}{self.name}"
        if not config.has_section(section):
            config.add_section(section)
        config.set(section, "url", self.url)
        config.set(section, "token", self.token or "")
        config.set(section, "type", self.auth_type)
        if self.project:
            config.set(section, "project", self.project)
        else:
            # a re-save without a project must not resurrect the old one
            config.remove_option(section, "project")


def contree_home() -> Path:
    home = os.environ.get("CONTREE_HOME")
    if home:
        return Path(home).expanduser()
    xdg = os.environ.get("XDG_CONFIG_HOME") or "~/.config"
    return Path(xdg).expanduser() / "contree"


def load_profiles(
    path: str | Path | None = None,
) -> tuple[dict[str, Profile], str]:
    """Read all profiles; return `(profiles, active_profile_name)`.

    `cli.ini` is read first, then `auth.ini` - values from `auth.ini`
    win on conflicts.
    """
    auth_file = Path(path) if path is not None else contree_home() / "auth.ini"
    parser = configparser.ConfigParser()
    parser.read([auth_file.parent / "cli.ini", auth_file])
    active = parser.defaults().get("profile", DEFAULT_PROFILE)
    profiles: dict[str, Profile] = {}
    for section in parser.sections():
        if not section.startswith(PROFILE_PREFIX):
            continue
        name = section[len(PROFILE_PREFIX) :]
        auth_type = parser.get(section, "type", fallback=AUTH_TYPE_JWT)
        default_url = DEFAULT_IAM_URL if auth_type == AUTH_TYPE_IAM else ""
        profiles[name] = Profile(
            name=name,
            token=parser.get(section, "token", fallback=None),
            url=parser.get(section, "url", fallback=default_url).rstrip("/"),
            auth_type=auth_type,
            project=parser.get(section, "project", fallback=None),
        )
    return profiles, active


def from_env() -> Profile | None:
    """Build a profile from the environment, bypassing config files.

    Reads the standard Contree variables:
    ``CONTREE_TOKEN`` (or ``NEBIUS_API_KEY``), ``CONTREE_URL`` and
    optionally ``CONTREE_PROJECT`` (or ``NEBIUS_AI_PROJECT``).
    Reports None unless both a token and a URL are present, so callers
    can fall through to :func:`resolve_profile`.
    """
    token = os.environ.get("CONTREE_TOKEN") or os.environ.get("NEBIUS_API_KEY")
    url = os.environ.get("CONTREE_URL")
    if not token or not url:
        return None
    return Profile(
        name="environment",
        url=url.rstrip("/"),
        token=token,
        project=os.environ.get("CONTREE_PROJECT")
        or os.environ.get("NEBIUS_AI_PROJECT"),
    )


def resolve_profile(
    name: str | None = None,
    *,
    path: str | Path | None = None,
) -> Profile:
    """Resolve the active profile by name.

    Priority: explicit *name* > `CONTREE_PROFILE` environment variable
    > the active profile recorded in the config file.
    """
    profiles, active = load_profiles(path)
    selected = name or os.environ.get("CONTREE_PROFILE") or active
    if selected not in profiles:
        raise ProfileError(f"profile {selected!r} not found")
    return profiles[selected]
